source/common/http/async_client_impl.cc

```diff
@@ -87,6 +90,18 @@ void AsyncStreamImpl::encodeHeaders(HeaderMapPtr&& headers, bool end_stream) {
 void AsyncStreamImpl::encodeData(Buffer::Instance& data, bool end_stream) {
   ENVOY_LOG(trace, "async http request response data (length={} end_stream={})", data.length(),
             end_stream);
+  // It's possible we have been reset by the client but the router is still
+  // sending data on its way out. This happens for example when
+  // Utility::sendLocalReply() tells the gRPC async client that it has a non-200
+  // via encodeHeaders(), gRPC async client resets the HTTP async stream but on
+  // return to Utility::sendLocalReply() it unconditionally sends a body with
+  // encodeData().
+  // TODO(htuch): This is the only case we catch today, but conceivably we could
+  // see encodeTrailers() as well (or encodeHeaders()?) from router on its way
```

Comment:
Per above I think we need to guard encodeTrailers() if we do the fix like this. I don't understand the comment about encodeHeaders() ? You guard that above, but I don't think that can happen?

In general, I agree this is sketchy. Arguably, this is a bug in Utility::secondLocalReply(). Basically, even local replies should adhere to a reset and stop sending data. Fixing that is probably out of scope for this change, but I might put in a TODO/NOTE that we should look at this more.

re: coverage you should be able to hit this via unit tests I think?